js-api-spec/compiler.test.ts

```diff
@@ -0,0 +1,174 @@
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import type {CompileResult, Importer} from 'sass';
+import {
+  initAsyncCompiler,
+  initCompiler,
+  SassString,
+  AsyncCompiler,
+  Compiler,
+} from 'sass';
+
+import {spy, URL} from './utils';
+
+export const functions = {
+  'foo($args)': (args: unknown) => new SassString(`${args}`),
+};
+
+export const importers: Array<Importer> = [
+  {
+    canonicalize: (url: string) => new URL(`u:${url}`),
+    load: (url: typeof URL) => ({
+      contents: `.import {value: ${url.pathname}} @debug "imported";`,
+      syntax: 'scss' as const,
+    }),
+  },
+];
+
+export const asyncImporters: Array<Importer> = [
+  {
+    canonicalize: (url: string) => Promise.resolve(new URL(`u:${url}`)),
+    load: (url: typeof URL) => Promise.resolve(importers[0].load(url)),
+  },
+];
+
+export const getLogger = () => ({debug: spy(() => {})});
+
+/* A trigged importer that executes a callback after a trigger is called */
+export function getTriggeredImporter(callback: () => void): {
+  importer: Importer;
+  triggerComplete: () => void;
+} {
+  let promiseResolve: (value: unknown) => void;
+  const awaitedPromise = new Promise(resolve => {
+    promiseResolve = resolve;
+  });
+  return {
+    importer: {
+      canonicalize: async () => new URL('foo:bar'),
+      load: async () => {
+        await awaitedPromise;
+        callback();
+        return {contents: '', syntax: 'scss' as const};
+      },
+    },
+    triggerComplete: () => promiseResolve(undefined),
+  };
+}
+
+describe('Compiler', () => {
+  let compiler: Compiler;
+
+  beforeEach(() => {
+    compiler = initCompiler();
+  });
+
+  afterEach(() => {
+    compiler.dispose();
+  });
+
+  describe('compileString', () => {
+    it('performs complete compilations', () => {
+      const logger = getLogger();
+      const result = compiler.compileString(
+        '@import "bar"; .fn {value: foo(baz)}',
+        {importers, functions, logger}
+      );
+      expect(result.css).toEqualIgnoringWhitespace(
+        '.import {value: bar;} .fn {value: "baz";}'
+      );
+      expect(logger.debug).toHaveBeenCalledTimes(1);
+    });
+
+    it('performs compilations in callbacks', () => {
+      const nestedImporter: Importer = {
+        canonicalize: () => new URL('foo:bar'),
+        load: () => ({
+          contents: compiler.compileString('x {y: z}').css,
+          syntax: 'scss',
+        }),
+      };
+      const result = compiler.compileString('@import "nested"; a {b: c}', {
+        importers: [nestedImporter],
+      });
+      expect(result.css).toEqualIgnoringWhitespace('x {y: z;} a {b: c;}');
+    });
+
+    it('throws after being disposed', () => {
+      compiler.dispose();
+      expect(() => compiler.compileString('$a: b; c {d: $a}')).toThrowError();
+    });
+  });
+  it('errors if constructor invoked directly', () => {
```

Comment:
Blank line before test case

Reply:
Added in https://github.com/sass/sass-spec/pull/1954/commits/76e676aaddc313aa5a53835cd0c883f54e530314